### PR TITLE
Lesson2 Denis Sulima

### PR DESCRIPTION
**Pull request by SD3nis**

### CODE REVIEW

File: Lesson1/Sulima Denis Homework 1/include/Utils.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#include <vector>
+
+
+
+
+namespace avprocessing
+{
+namespace utilities
+{
+    void reverse(std::forward_list<int> & m_ForvardList);
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/include/Utils.h]: Lesson1 is present. You didn't get the idea about branches.

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#include "functions.h"
+#include <stack>
+
+    bool isExpressionValid (const std::string& expression){
+        if(expression.empty())
+            return false;
+        std::stack<char>stackOpenBrackets;
+        for(size_t i = 0; i < expression.size(); i++){
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp]: std::string is pseudocontainer, so you can use the range for

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#include "functions.h"
+#include <stack>
+
+    bool isExpressionValid (const std::string& expression){
+        if(expression.empty())
+            return false;
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp]: _Why?_

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        std::stack<char>stackOpenBrackets;
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp]: **Try to use two stacks.**

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            if(expression[i] == '{' || expression[i] == '(' || expression[i] == '[')
+            {
+                stackOpenBrackets.push(expression[i]);
+            }
+            else {
+                if(stackOpenBrackets.empty()){
+                    return false;
+                }
+                char reverseRightToLeftBrackets;
+                switch (expression[i])
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp]: please extract to new function

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#include "functions.h"
+#include <stack>
+#include <queue>
+
+char reverseRightToLeftBracketsFunction(const char m_rightBrackets){
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp:5]: _argument naming_

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    case ')':
+    {
+        return '(';
+    }
+    case ']':
+    {
+        return '[';
+    }
+    default:
+        return NULL;
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp:21]: _exception?_

File: Lesson3/Sulima Homework 2/src/functions.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    if(stackOpenBrackets.empty()){
+        return true;
+    }
+    return false;
+}
+
+
+void linkLevelNodes(node *root)
+{
+    std::queue<node *> firstLevel;
```
Comment (smay1613) [inline Lesson3/Sulima Homework 2/src/functions.cpp:59]: what if root is nullptr?

File: Lesson3/Sulima Homework 2/src/functions.cpp (hunk 7 above)
Comment (SD3nis) [reply Lesson3/Sulima Homework 2/src/functions.cpp:21]: yes, if we pass a no bracket to the function, we get NULL.

**Pull request closed without merging**